internal/util/fetch/fetch.go

```diff
@@ -0,0 +1,284 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fetch
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt/internal/gitutil"
+	"github.com/GoogleContainerTools/kpt/internal/util/git"
+	"github.com/GoogleContainerTools/kpt/pkg/kptfile"
+	"github.com/GoogleContainerTools/kpt/pkg/kptfile/kptfileutil"
+	"sigs.k8s.io/kustomize/kyaml/copyutil"
+	"sigs.k8s.io/kustomize/kyaml/errors"
+)
+
+// Command fetches a package from a git repository and copies it to a local
+// directory.
+type Command struct {
+	RepoSpec *git.RepoSpec
+
+	Destination string
+
+	Name string
+
+	Clean bool
+}
+
+// Run runs the Command.
+func (c Command) Run() error {
+	if err := (&c).DefaultValues(); err != nil {
+		return err
+	}
+
+	if _, err := os.Stat(c.Destination); !c.Clean && !os.IsNotExist(err) {
+		return errors.Errorf("destination directory %s already exists", c.Destination)
+	}
+
+	// normalize path to a filepath
+	if !strings.HasSuffix(c.RepoSpec.Dir, "file://") {
+		c.RepoSpec.Dir = filepath.Join(path.Split(c.RepoSpec.Dir))
+	}
+
+	err := cloneAndCopy(c.RepoSpec, c.Destination, c.Name, c.Clean)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// cloneAndCopy fetches the provided repo and copies the content into the
+// directory specified by dest. The provided name is set as `metadata.name`
+// of the Kptfile of the package.
+func cloneAndCopy(r *git.RepoSpec, dest, name string, clean bool) error {
```

Comment:
This is a broader question about get. Should we have the packages fetched to some temp directory first on local, process everything including auto-setters and then only write it back to actual destination? That way, users wouldn't have to clean up the fetched package in case of errors ?

Reply:
So for the get command itself, I'm not sure if that helps us much. With subpackages we can still get into the situation where we have fetch/updated some packages and only then run into errors. 
But we could consider whether the complete get/update flow should happen in a temp directory and we only move it over at the end. It should mean that only errors during the final copy would require users to clean up manually. I think we should consider that separately once we have the main features implemented.